### ISSUE TITLE
Youtube URL Decryption Fails

### ISSUE THREAD

**Issue by The-Demon-Prosecutor** [TeamNewPipe/NewPipeExtractor#133]
Youtube has changed something in their code base again which results in NewPipe unable to decrypt the URL.

I would like to point out that youtube-dl has a fix already [Khang-NT@f829ef2](https://github.com/Khang-NT/youtube-dl/commit/f829ef2e13c4e791f455282a9b7f0ecde9891252).

Thanks

Edit: Paging @karyogamy

**Comment by atlanntiss** [comment]
The same problem.

**Comment by delightfulagony** [comment]
Same problem here, here's the log:
```
{
   "user_action": "requested stream",
   "request": "https:\/\/www.youtube.com\/watch?v=7pDfWhGN3y4",
   "content_language": "ES",
   "service": "YouTube",
   "package": "org.schabi.newpipe",
   "version": "0.14.2",
   "os": "Linux Android 7.1.2 - 25",
   "time": "2019-01-17 01:34",
   "exceptions": [
      "org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor$DecryptException: Could not parse decrypt function \n\tat org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor.loadDecryptionCode(YoutubeStreamExtractor.java:729)\n\tat org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor.onFetchPage(YoutubeStreamExtractor.java:604)\n\tat org.schabi.newpipe.extractor.Extractor.fetchPage(Extractor.java:51)\n\tat org.schabi.newpipe.extractor.stream.StreamInfo.getInfo(StreamInfo.java:60)\n\tat org.schabi.newpipe.extractor.stream.StreamInfo.getInfo(StreamInfo.java:56)\n\tat org.schabi.newpipe.util.ExtractorHelper.lambda$getStreamInfo$3(ExtractorHelper.java:109)\n\tat org.schabi.newpipe.util.-$$Lambda$ExtractorHelper$5fJcha6Sq5APJBLdG6osaJby-mc.call(lambda)\n\tat io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)\n\tat io.reactivex.Single.subscribe(Single.java:3438)\n\tat io.reactivex.internal.operators.single.SingleDoOnSuccess.subscribeActual(SingleDoOnSuccess.java:35)\n\tat io.reactivex.Single.subscribe(Single.java:3438)\n\tat io.reactivex.internal.operators.maybe.MaybeFromSingle.subscribeActual(MaybeFromSingle.java:41)\n\tat io.reactivex.Maybe.subscribe(Maybe.java:4154)\n\tat io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.drain(MaybeConcatArray.java:153)\n\tat io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.request(MaybeConcatArray.java:78)\n\tat io.reactivex.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onSubscribe(FlowableElementAtMaybe.java:66)\n\tat io.reactivex.internal.operators.maybe.MaybeConcatArray.subscribeActual(MaybeConcatArray.java:42)\n\tat io.reactivex.Flowable.subscribe(Flowable.java:14479)\n\tat io.reactivex.internal.operators.flowable.FlowableElementAtMaybe.subscribeActual(FlowableElementAtMaybe.java:36)\n\tat io.reactivex.Maybe.subscribe(Maybe.java:4154)\n\tat io.reactivex.internal.operators.maybe.MaybeToSingle.subscribeActual(MaybeToSingle.java:46)\n\tat io.reactivex.Single.subscribe(Single.java:3438)\n\tat io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)\n\tat io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)\n\tat io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)\n\tat io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:237)\n\tat java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)\n\tat java.lang.Thread.run(Thread.java:761)\nCaused by: org.schabi.newpipe.extractor.utils.Parser$RegexException: failed to find pattern \"(encodeURIComponent=function\\([a-zA-Z0-9_]+\\)\\{.+?\\})\n\tat org.schabi.newpipe.extractor.utils.Parser.matchGroup(Parser.java:63)\n\tat org.schabi.newpipe.extractor.utils.Parser.matchGroup1(Parser.java:52)\n\tat org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor.loadDecryptionCode(YoutubeStreamExtractor.java:713)\n\t... 30 more\n"
   ],
   "user_comment": ""
}
```
Although I'm sure it won't be difficult to replicate.

**Comment by fipsy1** [comment]
Same problems here :-(

**Comment by theScrabi** [comment]
Noticed. Could the fix be added to NewPipe extractor as well?

**Comment by theScrabi** [comment]
You can find the related code here:
https://github.com/TeamNewPipe/NewPipeExtractor/blob/master/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java#L568

**Comment by fabiano-araujo** [comment]
Same error

**Comment by Rub-z** [comment]
@thescabi just a thought but have you considered adding this DECYRYPTION_SIGNATURE_FUNCTION_REGEX string as an optional editable value on the settings page? Not only will it stop NewPipe breaking until a new release but it will also give the community a chance to help solve the Google riddle once more too! 

I'm not a java coding cowboy but surely this is not too difficult to implement? Just an idea.

**Comment by jdertl** [comment]
A basic solution to completely go around the issue would be:
1)  Change 'DECYRYPTION_SIGNATURE_FUNCTION_REGEX' to not be final, but provide the latest by default
2)  Give 'YoutubeStreamExtractor' optional save and load methods to read in and write out the 'DECYRYPTION_SIGNATURE_FUNCTION_REGEX' value
3)  Store the latest value for this in a separate file in the GitHub, or to the actual location where it could be copied from
4)  Upon explicit failure that causes this error, attempt to parse the file at the URL for that separate file (or location of where it can be copied from) and on success replace the value of 'DECYRYPTION_SIGNATURE_FUNCTION_REGEX'.

This way any service or app that uses the Extractor can simply load and save the previous regex from a file without constantly pinging out to GitHub or a place where it can be copied from, and only attempt to on failure.  This would also make it so in the worst case a single push of a single file would have to be made to fix everyone's Extractor (which could be optional, could be locked from happening with an option if people didn't want to have to manually cut out the code or similar), and in best case the fix is automated by simply copying it from the usable source where the regex could be found.

**Comment by The-Demon-Prosecutor** [comment]
We could use [youtube-dl's new regex](https://github.com/rg3/youtube-dl/commit/fa4ac365f69cbd51e4c9801984ebea49a12825b7) as a hack-y fix and migrating to the set-up as described by @jdertl for the long term.

**Comment by theScrabi** [comment]
> Not only will it stop NewPipe breaking until a new release but it will also give the community a chance to help solve the Google riddle once more too!

We can think about that :)

Anyway, I'm having my finals today, I cant help before finishing them, so I'd really appreciate if one could help and send a PR.

**Comment by B0rlaug** [comment]
Looks like the addition of "encodeURIComponent" fixes it, currently at work so I'll try to send a PR later today if no-one has already done so.
@theScrabi good luck on your finals man, will try to send it after work.

**Comment by The-Twisted-Samurai** [comment]
Likewise @B0rlaug. I seem to fail the getRelatedVideos test though. Experiencing the same?

**Comment by The-Twisted-Samurai** [comment]
PR is live!

**Comment by theScrabi** [comment]
I'm not FUCKING able to import that FUCKING gralde project on my FUCKING new computer!!!
I can't test this shit.

**Comment by theScrabi** [comment]
@The-Twisted-Samurai thx :)

**Comment by theScrabi** [comment]
This:
![failing_shit](https://user-images.githubusercontent.com/1891273/51384095-89b63f00-1b1b-11e9-9662-8a7ae9e27af8.png)

**Comment by The-Twisted-Samurai** [comment]
You mean the one in docs? Yes, I think I did do that.

**Comment by theScrabi** [comment]
That was the point, IDEA is not showing me. It sais tests are failing, but does not tell which. However it seems to work now :)

I am going to push the fixed version of NP later.

**Comment by The-Twisted-Samurai** [comment]
Ah I see. The curse of technology, eh? :p

**Comment by The-Twisted-Samurai** [comment]
Thank you for your work, @theScrabi. Best of luck with your finals!

**Comment by theScrabi** [comment]
:D thank you.

**Comment by jharbig** [comment]
Is the fix already in the new F-Droid build from yesterday (version 18.1)? I updated to it, but the problem still persists.

`{
   "user_action": "requested stream",
   "request": "https:\/\/www.youtube.com\/watch?v=Q1A7i7fHuew",
   "content_language": "DE",
   "service": "YouTube",
   "package": "org.schabi.newpipe",
   "version": "0.18.1",
   "os": "Linux Android 8.1.0 - 27",
   "time": "2020-01-24 10:23",
   "exceptions": [
      "org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor$DecryptException: Could not parse decrypt function \n\tat org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor.loadDecryptionCode(YoutubeStreamExtractor.java:849)\n\tat org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor.onFetchPage(YoutubeStreamExtractor.java:716)\n\tat org.schabi.newpipe.extractor.Extractor.fetchPage(Extractor.java:54)\n\tat org.schabi.newpipe.extractor.stream.StreamInfo.getInfo(StreamInfo.java:64)\n\tat org.schabi.newpipe.extractor.stream.StreamInfo.getInfo(StreamInfo.java:60)\n\tat org.schabi.newpipe.util.ExtractorHelper.lambda$getStreamInfo$3(ExtractorHelper.java:115)\n\tat org.schabi.newpipe.util.-$$Lambda$ExtractorHelper$5fJcha6Sq5APJBLdG6osaJby-mc.call(Unknown Source:4)\n\tat io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)\n\tat io.reactivex.Single.subscribe(Single.java:3438)\n\tat io.reactivex.internal.operators.single.SingleDoOnSuccess.subscribeActual(SingleDoOnSuccess.java:35)\n\tat io.reactivex.Single.subscribe(Single.java:3438)\n\tat io.reactivex.internal.operators.maybe.MaybeFromSingle.subscribeActual(MaybeFromSingle.java:41)\n\tat io.reactivex.Maybe.subscribe(Maybe.java:4154)\n\tat io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.drain(MaybeConcatArray.java:153)\n\tat io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.request(MaybeConcatArray.java:78)\n\tat io.reactivex.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onSubscribe(FlowableElementAtMaybe.java:66)\n\tat io.reactivex.internal.operators.maybe.MaybeConcatArray.subscribeActual(MaybeConcatArray.java:42)\n\tat io.reactivex.Flowable.subscribe(Flowable.java:14479)\n\tat io.reactivex.internal.operators.flowable.FlowableElementAtMaybe.subscribeActual(FlowableElementAtMaybe.java:36)\n\tat io.reactivex.Maybe.subscribe(Maybe.java:4154)\n\tat io.reactivex.internal.operators.maybe.MaybeToSingle.subscribeActual(MaybeToSingle.java:46)\n\tat io.reactivex.Single.subscribe(Single.java:3438)\n\tat io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)\n\tat io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)\n\tat io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)\n\tat io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)\n\tat java.lang.Thread.run(Thread.java:764)\nCaused by: org.schabi.newpipe.extractor.utils.Parser$RegexException: failed to find pattern \";([A-Za-z0-9_\\$]{2})\\...\\( inside of var lea=function(a){a=a.split(\"\");var b=[function(c,d){d=(d%c.length+c.length)%c.length;c.splice(-d).reverse().forEach(function(e){return c.unshift(e)};\"\n\tat org.schabi.newpipe.extractor.utils.Parser.matchGroup(Parser.java:73)\n\tat org.schabi.newpipe.extractor.utils.Parser.matchGroup(Parser.java:61)\n\tat org.schabi.newpipe.extractor.utils.Parser.matchGroup1(Parser.java:52)\n\tat org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor.loadDecryptionCode(YoutubeStreamExtractor.java:836)\n\t... 30 more\n"
   ],
   "user_comment": ""
}`

**Comment by The-Twisted-Samurai** [comment]
Uhh... this was a year ago. The current fix is in v0.18.2. See here [1][2].

References:
[1]: https://github.com/TeamNewPipe/NewPipe/releases
[2]: https://github.com/TeamNewPipe/NewPipeExtractor/pull/243
